pennylane/ops/qubit.py

```diff
@@ -2165,6 +2165,39 @@ def _controlled(self, wire):
         ControlledQubitUnitary(*self.parameters, control_wires=wire, wires=self.wires)
 
 
+class QubitMatrix(Operation):
+    r"""QubitMatrix(A, wires)
+    Apply an arbitrary square matrix.
```

Comment:
I'm tempted to say, let's not have any constraints? It is up to the device if it wants to support this operation, which may be non-physical? I don't know 🤔

Reply:
Sure, I like the idea of keeping constraints to a minimum and trying to keep this operator as flexible as possible. Would certainly be worth testing the edge cases, happy to do that!

Reply:
> It is up to the device if it wants to support this operation, which may be non-physical?

I was wondering what the difference was with this and `QubitUnitary` - so basically this can be *any* square matrix, not necessarily unitary? What's an example use case?

Reply:
I naturally came across a use case when trying to output the density matrix of a circuit after a projection operation:

```
eigenstate = 2**(-.5)*np.array([1, 1j])
P = np.outer(eplus_i, np.conjugate(eigenstate))

@qml.qnode(dev)
def circuit():
    
    qml.Hadamard(wires=0)
    qml.CNOT(wires=[0,1])
    
    qml.QubitOperator(P, wires=1)
    
    return qml.density_matrix(wires)
```

Though this example is for a Hermitian operator and could potentially be solved by allowing `qml.Hermitian()` to inherit from the `Operator` `class`; it currently inherits from `Obsevable` only. But since non-Hermitian operators can be used in circuits (e.g `S`, `T`, `P`) it may be useful to keep it more general.

Reply:
Also, this could make state preparation very simple for users unfamiliar with standard gate operations